sklearn/preprocessing/_discretization.py

```diff
@@ -225,13 +237,17 @@ def fit(self, X, y=None):
                         n_samples, size=self.subsample, replace=False
                     )
                     X = _safe_indexing(X, subsample_idx)
-        elif self.strategy != "quantile" and isinstance(
-            self.subsample, numbers.Integral
-        ):
-            raise ValueError(
-                f"Invalid parameter for `strategy`: {self.strategy}. "
-                '`subsample` must be used with `strategy="quantile"`.'
-            )
+        elif self.strategy != "quantile":
+            if isinstance(self.subsample, numbers.Integral):
+                raise ValueError(
+                    f"Invalid parameter for `strategy`: {self.strategy}. "
+                    '`subsample` must be used with `strategy="quantile"`.'
+                )
+            if sample_weight is not None:
```

Comment:
What I mean is to isolate this case in a later statement

```python
if sample_weight is not None and self.strategy != "quantile":
   raise ...
```

In this way is that we are validating `sample_weight`. Right now we validate 2 things (i.e. `subsample` and `sample_weight` which is a bit more complex to understand.

Reply:
I isolated this check in its own `if` statement (l. 247). https://github.com/scikit-learn/scikit-learn/pull/22048/commits/1a417f4d1d6ee00213cd576bd41e407fc45c88e5